ipa-core/src/protocol/ipa_prf/quicksort.rs

```diff
@@ -0,0 +1,210 @@
+/// To see why this leaks information: assume that all elements are equal,
+/// then quicksort runs in time O(n^2) while for unique elements, it is only expected to run in time O(n log n)
+///
+/// The leakage can be fixed by appending a counter on each element that is unique to the element
+/// This adds another >= log N bits, where N is the amount of elements
+///
+/// This implementation of quicksort is in place and uses a stack instead of recursion
+/// It terminates once the stack is empty
+/// # Errors
+/// Will propagate errors from transport and a few typecasts
+#[cfg(all(test, unit_test))]
+pub async fn quicksort_by_key_insecure<C, K, F, S>(
+    ctx: C,
+    list: &mut [S],
+    desc: bool,
+    f: F,
+) -> Result<(), Error>
+where
+    C: Context,
+    S: Send + std::marker::Sync,
+    F: FnMut(&S) -> &AdditiveShare<K> + std::marker::Send + Clone + std::marker::Sync,
+    for<'a> &'a AdditiveShare<K>: IntoIterator<Item = AdditiveShare<K::Element>>,
+    K: WeakSharedValue + Field + CustomArray<Element = Boolean>,
+{
+    // create stack
+    let mut stack: Vec<(C, usize, usize)> = vec![];
+
+    // initialize stack
+    stack.push((ctx, 0usize, list.len()));
+
+    // iterate through quicksort recursions
+    while let Some((ctx, b_l, b_r)) = stack.pop() {
+        // start of quicksort function
+        // check whether sort is needed
+        if b_l + 1 < b_r {
+            // set up iterator
+            let mut iterator = list[b_l..b_r].iter();
+            // first element is pivot, apply key extraction function f
+            let pivot = f.clone()(iterator.next().unwrap());
+            // create pointer to context for moving into closure
+            let pctx = &ctx;
+            // precompute comparison against pivot and reveal result in parallel
+            let comp = seq_join(
+                ctx.active_work(),
+                stream_iter(iterator.enumerate().map(|(n, x)| {
+                    let mut fc = f.clone();
+                    async move {
+                        // compare current element against pivot
+                        let sh_comp = compare_gt(
+                            pctx.narrow(&Step::Compare)
+                                .set_total_records(b_r - (b_l + 1)),
+                            RecordId::from(n),
+                            // apply key extraction function f to element x
+                            fc(x),
+                            pivot,
+                        )
+                        .await?;
+
+                        // reveal outcome of comparison
+                        sh_comp
+                            .reveal(
+                                pctx.narrow(&Step::Reveal)
+                                    .set_total_records(b_r - (b_l + 1)),
+                                RecordId::from(n),
+                            )
+                            .await
+                    }
+                })),
+            )
+            .try_collect::<Vec<_>>()
```

Comment:
there is likely an optimization to keep those bits inside a bitvec and/or reuse it between iterations

Reply:
Using a bitvec here by unwrapping Boolean(bool) would make sense to save some memory costs. We would need to generate bitvec with lengths that are only known at run time. We wouldn't save any communication costs since these bits are not used in any MPC computation. Further, these bits are not reused across iterations but we could store them in the same memory location.